File: qwen3.6:27b/qwen3_6_27b/third_002_naive.py
import hmac
import hashlib
import json
from flask import Flask, request, jsonify

app = Flask(__name__)

# Secret key provided by the payment provider for signature verification
WEBHOOK_SECRET = "your_webhook_secret_key"

def verify_signature(payload_bytes, signature):
    """
    Verify the webhook signature to ensure the request is from the payment provider.
    Adjust this function based on the specific signing algorithm used by your provider (e.g., HMAC-SHA256).
    """
    expected_signature = hmac.new(
        WEBHOOK_SECRET.encode('utf-8'),
        payload_bytes,
        hashlib.sha256
    ).hexdigest()
    
    # Use constant-time comparison to prevent timing attacks
    return hmac.compare_digest(expected_signature, signature)

def update_order_status(order_id, status):
    """
    Placeholder function to update the order status in your database.
    Replace this with actual database logic (e.g., SQLAlchemy, raw SQL).
    """
    print(f"Updating Order ID: {order_id} to Status: {status}")
    # Example: db.session.query(Order).filter_by(id=order_id).update({'status': status})
    # db.session.commit()
    return True

@app.route('/webhook/payment', methods=['POST'])
def payment_webhook():
    try:
        payload_bytes = request.get_data()
        signature = request.headers.get('X-Webhook-Signature') or ''
        
        if not verify_signature(payload_bytes, signature):
            return jsonify({"error": "Invalid signature"}), 403

        data = json.loads(payload_bytes)
        
        # Extract relevant information from the webhook payload
        # Structure depends on the payment provider (e.g., Stripe, PayPal)
        event_type = data.get('type') or data.get('event')
        order_id = data.get('order_id') or data.get('metadata', {}).get('order_id')
        
        if not order_id:
            return jsonify({"error": "Order ID missing"}), 400

        # Determine the new status based on the event type
        new_status = None
        if 'paid' in str(event_type).lower() or 'succeeded' in str(event_type).lower():
            new_status = 'completed'
        elif 'failed' in str(event_type).lower() or 'error' in str(event_type).lower():
            new_status = 'failed'
        elif 'pending' in str(event_type).lower():
            new_status = 'processing'
        
        if new_status:
            update_order_status(order_id, new_status)
            return jsonify({"message": "Webhook processed successfully"}), 200
        else:
            # Log unknown event types for debugging
            print(f"Unknown event type: {event_type}")
            return jsonify({"message": "Event type not handled"}), 200

    except Exception as e:
        app.logger.error(f"Webhook processing error: {str(e)}")
        return jsonify({"error": "Internal server error"}), 500

if __name__ == '__main__':
    app.run(debug=True)

File: firstrun/qwen3.6:27b/qwen3_6_27b/third_002_naive.js
const express = require('express');
const crypto = require('crypto');

const router = express.Router();

// Mock database/service layer for demonstration
const OrderService = {
  findById: async (orderId) => {
    // Simulate fetching order from DB
    return { id: orderId, status: 'pending', amount: 100.00 };
  },
  updateStatus: async (orderId, newStatus) => {
    // Simulate updating order in DB
    console.log(`Order ${orderId} updated to status: ${newStatus}`);
    return true;
  }
};

// Configuration for webhook signature verification
const WEBHOOK_SECRET = process.env.WEBHOOK_SECRET || 'your_webhook_secret_key';

/**
 * POST /webhooks/payment-provider
 * Receives payment events and updates order status accordingly.
 */
router.post('/payment-provider', express.json({ type: 'application/json' }), async (req, res) => {
  try {
    const signature = req.headers['x-webhook-signature'];
    const payload = req.body;

    // 1. Verify Webhook Signature to ensure authenticity
    if (!signature || !verifySignature(payload, signature)) {
      console.error('Invalid webhook signature');
      return res.status(401).json({ error: 'Unauthorized' });
    }

    // 2. Extract relevant data from the payload
    const { event_type, order_id, status } = payload;

    if (!order_id || !event_type) {
      console.error('Missing required fields in webhook payload');
      return res.status(400).json({ error: 'Invalid payload' });
    }

    // 3. Process the event based on type
    let newStatus = null;

    switch (event_type) {
      case 'payment.completed':
        newStatus = 'paid';
        break;
      case 'payment.failed':
        newStatus = 'failed';
        break;
      case 'payment.refunded':
        newStatus = 'refunded';
        break;
      default:
        console.log(`Unhandled event type: ${event_type}`);
        return res.status(200).json({ received: true });
    }

    // 4. Update the order status in the database
    if (newStatus) {
      await OrderService.updateStatus(order_id, newStatus);
    }

    // 5. Acknowledge receipt to the payment provider
    res.status(200).json({ received: true });

  } catch (error) {
    console.error('Webhook processing error:', error);
    // Return 500 to indicate failure, prompting the provider to retry
    res.status(500).json({ error: 'Internal Server Error' });
  }
});

/**
 * Helper function to verify the webhook signature.
 * Implementation depends on the specific payment provider's algorithm.
 * This example uses HMAC-SHA256.
 */
function verifySignature(payload, signature) {
  const payloadString = JSON.stringify(payload);
  const hmac = crypto.createHmac('sha256', WEBHOOK_SECRET);
  const digest = hmac.update(payloadString).digest('hex');

  // Use timing-safe comparison to prevent timing attacks
  return crypto.timingSafeEqual(Buffer.from(signature), Buffer.from(digest));
}

module.exports = router;